src/v/raft/state_machine.cc

```diff
@@ -105,10 +113,16 @@ ss::future<> state_machine::apply() {
     return _raft->events()
       .wait(_next, model::no_timeout, _as)
       .then([this] {
-          // build a reader for log range [_next, +inf).
-          storage::log_reader_config config(
-            _next, model::model_limits<model::offset>::max(), _io_prio);
-          return _raft->make_reader(config);
+          auto f = ss::now();
+          if (_next < _raft->start_offset()) {
+              f = handle_eviction();
```

Comment:
maybe it would be good to update `_next` here. this way we will not have to rely on somebody implementing `handle_log_eviction` to do that ?

Reply:
Some thoughts on why that would be undesirable: if the state machine handles eviction by loading the snapshot, we would want it to set `_next` based on the snapshot header. If this is inconsistent with `_raft->start_offset()`, we would at least notice.